STORM-2371: Replace existing AMQP eventhub client with the lastest one from Microsoft eventhubs

rban1 opened the pull request:
Incorporating latest eventhub API changes

SreeramGarlapati commented on external/storm-eventhubs/src/main/java/org/apache/storm/eventhubs/spout/StringEventDataScheme.java:
```diff
@@ -44,21 +40,23 @@
+      messageData = new String (eventData.getBytes());
+    else if(eventData.getObject()!=null){
+      try{
+        messageData = new String(Serializedeserializeutil.serialize(eventData.getObject()),Charset.defaultCharset());
```
>new String(Serializedeserializeutil.serialize(eventData.getObject()),Charset.defaultCharset()); [](start = 22, length = 95)

eventData.getObject() would return java type of the corresponding Amqp object (if there is no direct java type mapping available - you would see a proton-j type here).
If the type is a primitive data type - simply return .tostring().
else - follow the serialization approach.

SreeramGarlapati commented on external/storm-eventhubs/src/main/java/org/apache/storm/eventhubs/spout/StringEventDataScheme.java:
```diff
@@ -44,21 +40,23 @@
+      try{
+        messageData = new String(Serializedeserializeutil.serialize(eventData.getObject()),Charset.defaultCharset());
+      }catch (IOException e){
+        logger.error("Failed to serialize object"+e.toString());
```
>object [](start = 42, length = 6)

nit: i would pitch the msg a bit more specific, like - failed to Serialize EventData payload

SreeramGarlapati commented on external/storm-eventhubs/src/main/java/org/apache/storm/eventhubs/spout/BinaryEventDataScheme.java:
```diff
@@ -37,23 +37,21 @@
+			try {
+				messageData = Serializedeserializeutil.serialize(eventData.getObject());
+			}catch (IOException e){
+				logger.error("Failed to serialize object"+e.toString());
```
>logger.error("Failed to serialize object"+e.toString()); [](start = 4, length = 56)

Will this not result into message loss..? if so don't Swallow the exception...

SreeramGarlapati commented on external/storm-eventhubs/src/main/java/org/apache/storm/eventhubs/spout/Serializedeserializeutil.java:
```diff
@@ -0,0 +1,21 @@
+
+/**
+ * Created by rabaner on 3/22/2017.
+ */
```
nit: license header and fix name of class to camel casing

SreeramGarlapati commented on pom.xml:
```diff
@@ -297,6 +297,7 @@
         <wagonVersion>1.0</wagonVersion>
         <qpid.version>0.32</qpid.version>
+        <azure-eventhubs.version>0.13.0</azure-eventhubs.version>
         <eventhubs.client.version>1.0.1</eventhubs.client.version>
```
><eventhubs.client.version>1.0.1</eventhubs.client.version> [](start = 8, length = 58)

which client is this ?

SreeramGarlapati commented on external/storm-eventhubs/src/main/java/org/apache/storm/eventhubs/spout/EventHubReceiverImpl.java:
```diff
@@ -65,77 +57,81 @@ public EventHubReceiverImpl(EventHubSpoutConfig config, String partitionId) {
+    Iterable<EventData> receivedEvents=null;
+    /*Get one message at a time for backward compatibility behaviour*/
+    try {
+      receivedEvents = receiver.receiveSync(1);
```
>receiver [](start = 23, length = 8)

this will result in NPE if in the Open() - createReceiver call threw exception (and the try catch swallowed it)

rban1 replied:
Made the change

SreeramGarlapati commented on external/storm-eventhubs/src/main/java/org/apache/storm/eventhubs/spout/SimplePartitionManager.java:
```diff
@@ -65,16 +60,13 @@ public void open() throws Exception {
-    }
-    else {
-      filter = new EventHubOffsetFilter(offset);
+      offset = Long.toString(config.getEnqueueTimeFilter());
```
>getEnqueueTimeFilter [](start = 36, length = 20)

why is TimeFilter translated to Offset filter ?

SreeramGarlapati commented on external/storm-eventhubs/src/main/java/org/apache/storm/eventhubs/spout/EventDataScheme.java:
```diff
@@ -44,27 +43,21 @@
 		}
-
+		}
+		Map metaDataMap = eventData.getProperties();
```
>Map metaDataMap = eventData.getProperties(); [](start = 2, length = 44)

the old solution is not optimal - will result into too many unnecessary map objects initialized; pls evaluate if you want to change it
if map.count()==0 (ie., empty map) - dont even return them...

harshach commented:
@rban1 you are keep re-opening the PRs. We should keep only one PR. If you want to address the comments and update the PR , all you need to do is to work on the same git branch and push changes into the same branch that will update the PR you opened.

rban1 commented:
@SreeramGarlapati  Made all the relavant changes(Serialization, exception handling, enqueuetimefilter)
@harshach I will be using this PR going forward. I have addressed your comments on syntax as well. On your comment of backporting the changes to 1.x branches it cannot be done as the latest eventhubclient is valid only from java 8 environment.

SreeramGarlapati commented on external/storm-eventhubs/src/main/java/org/apache/storm/eventhubs/spout/EventDataScheme.java:
```diff
@@ -44,27 +41,31 @@
-
+		Map metaDataMap = eventData.getProperties().size() > 0 ? eventData.getProperties() : null;
 		fieldContents.add(messageData);
 		fieldContents.add(metaDataMap);
```
>fieldContents.add(metaDataMap); [](start = 2, length = 31)

Do you still have to add this if the properties are null ?

rban1 replied:
will remove them

SreeramGarlapati commented on external/storm-eventhubs/src/main/java/org/apache/storm/eventhubs/spout/PartitionManager.java:
```diff
@@ -41,29 +41,29 @@ public PartitionManager(
-    this.pending = new LinkedHashMap<String, EventData>();
-    this.toResend = new TreeSet<EventData>();
+    this.pending = new LinkedHashMap<String, EventDataWrap>();
+    this.toResend = new TreeSet<EventDataWrap>();
```
>() [](start = 46, length = 2)

order by sequence number

rban1 replied:
it is indeed ordered by sequence number. The compareto definition in eventdatawrap class is as follows:
@Override
  public int compareTo(EventDataWrap ed) {
    return messageId.getSequenceNumber().
        compareTo(ed.getMessageId().getSequenceNumber());
  }

SreeramGarlapati commented on external/storm-eventhubs/src/main/java/org/apache/storm/eventhubs/spout/SerializeDeserializeUtil.java:
```diff
@@ -0,0 +1,35 @@
+        try (ByteArrayOutputStream b = new ByteArrayOutputStream()) {
+            try (ObjectOutputStream o = new ObjectOutputStream(b)) {
+                o.writeObject(obj);
+                o.close();
```
>o.close(); [](start = 16, length = 10)

not needed...

SreeramGarlapati commented on external/storm-eventhubs/src/main/java/org/apache/storm/eventhubs/spout/StringEventDataScheme.java:
```diff
@@ -44,21 +38,32 @@
+        if (!(eventData.getObject() instanceof List)) {
+          messageData = eventData.getObject().toString();
+        } else {
+          throw new RuntimeException("Cannot serialize the given AMQP type.");
```
>serialize [](start = 45, length = 9)

cannot deserialize...

rban1 replied:
will change it

SreeramGarlapati commented:
:shipit:

harshach commented:
@rban1 can you change the PR title to reflect the JIRA title. You can look other PRs for example

rban1 commented:
Made the necessary point changes and updated pom to use 0.13.1
@harshach  I have changed the PR name appropriately. Also rebased and I have only one commit in the commit history.

harshach commented:
+1
